openstack/03a-IPHC/frag.c

```diff
@@ -0,0 +1,756 @@
+
+//=========================== prototypes ======================================
+
+static void cleanup_fragments(uint16_t datagram_tag);
+
+static void store_fragment(OpenQueueEntry_t *msg, uint16_t size, uint16_t tag, uint8_t offset);
+
+static void reassemble_fragments(uint16_t tag, uint16_t size, OpenQueueEntry_t *reassembled_msg);
+
+static owerror_t allocate_vrb(OpenQueueEntry_t *frag1, uint16_t size, uint16_t tag);
+
+static void prepend_frag1_header(OpenQueueEntry_t *frag1, uint16_t size, uint16_t tag);
+
+static void prepend_fragn_header(OpenQueueEntry_t *fragn, uint16_t size, uint16_t tag, uint8_t offset);
+
+static void fast_forward_frags(uint16_t tag, uint16_t size, uint8_t vrb_pos);
+
+void frag_timeout_cb(opentimers_id_t id);
+
+owerror_t frag_timerq_enqueue(opentimers_id_t id);
+
+owerror_t frag_timerq_remove(opentimers_id_t id);
+
+opentimers_id_t frag_timerq_dequeue(void);
+//============================= public ========================================
+
+void frag_init() {
+    memset(&frag_vars, 0, sizeof(frag_vars_t));
+
+    // unspecified start value, wraps around at 65535     
+    frag_vars.global_tag = openrandom_get16b() & 0x7FF;
+}
+
+owerror_t frag_fragment6LoPacket(OpenQueueEntry_t *msg) {
+    OpenQueueEntry_t *lowpan_fragment;
+    uint16_t remaining_bytes;
+    uint8_t fragment_length; uint8_t fragment_offset;
+    int8_t bpos;
+
+    // check if fragmentation is necessary
+    if (!msg->l3_isFragment && msg->length > (MAX_FRAGMENT_SIZE + FRAGN_HEADER_SIZE)) {
+
+        openserial_printInfo(COMPONENT_FRAG, ERR_FRAG_FRAGMENTING,
+                              (errorparameter_t) msg->length,
+                              (errorparameter_t) (msg->length / MAX_FRAGMENT_SIZE) + 1);
+
+        // update the global 6LoWPAN datagram tag
+        frag_vars.global_tag++;
+        remaining_bytes = msg->length;
+        fragment_offset = 0;
+        bpos = -1;
+
+        // start fragmenting
+        while (remaining_bytes > 0) {
+
+            lowpan_fragment = openqueue_getFreePacketBuffer(COMPONENT_FRAG);
+            if (lowpan_fragment == NULL) {
+                // 6LoWPAN packet couldn't be entirely fragmented, thus clean up previously created fragments and exit.
+                openserial_printError(COMPONENT_FRAG, ERR_NO_FREE_PACKET_BUFFER,
+                        (errorparameter_t) 0,
+                        (errorparameter_t) 0);
+                cleanup_fragments(frag_vars.global_tag);
+                return E_FAIL;
+            }
+
+            lowpan_fragment->l3_isFragment = TRUE;
+            lowpan_fragment->owner = COMPONENT_FRAG;
+            lowpan_fragment->creator = msg->creator;
+
+            if (remaining_bytes > MAX_FRAGMENT_SIZE)
+                fragment_length = MAX_FRAGMENT_SIZE;
+            else
+                fragment_length = remaining_bytes;
+
+            // find a new spot in the fragmentation buffer  
+            for (int i = 0; i < FRAGMENT_BUFFER_SIZE; i++) {
```

Comment:
need to replace int with 'stdint.h definition'.

I prefer to define i in advance, rather than define it in side the for loop, like here. 
You may reuse the 'i' latter.